File: backend/seed.py
"""
Database Seed Script for Lonaat
Creates test users for development and acceptance testing
"""

from models import (
    db,
    User,
    Product,
    AffiliateClick,
    Commission,
    WithdrawalRequest,
    Transaction,
    CommissionStatus,
    WithdrawalStatus,
    TransactionType,
    TransactionStatus,
)
from flask import Flask
from config import Config
import sys
from decimal import Decimal
from datetime import timedelta


def create_deterministic_seed(app):
    """Create deterministic seed data:
    - 5 users (1 admin, 4 affiliates)
    - 10 offers (products)
    - 50 clicks
    - 12 commission/conversion records (8 approved/paid)
    - 3 completed payouts (withdrawal requests)

    The function is idempotent and will skip existing records by unique keys.
    """

    with app.app_context():
        print("\n🌱 Running deterministic seed...")

        # 1) Users
        users_spec = [
            {'name': 'Admin User', 'email': 'admin@lonaat.com', 'password': 'Admin123!', 'role': 'admin', 'is_admin': True, 'balance': Decimal('0')},
            {'name': 'Affiliate A', 'email': 'aff1@lonaat.com', 'password': 'AffPass1!', 'role': 'user', 'is_admin': False, 'balance': Decimal('100.00')},
            {'name': 'Affiliate B', 'email': 'aff2@lonaat.com', 'password': 'AffPass2!', 'role': 'user', 'is_admin': False, 'balance': Decimal('200.00')},
            {'name': 'Affiliate C', 'email': 'aff3@lonaat.com', 'password': 'AffPass3!', 'role': 'user', 'is_admin': False, 'balance': Decimal('300.00')},
            {'name': 'Affiliate D', 'email': 'aff4@lonaat.com', 'password': 'AffPass4!', 'role': 'user', 'is_admin': False, 'balance': Decimal('400.00')},
        ]

        created_users = []
        for u in users_spec:
            existing = User.query.filter_by(email=u['email']).first()
            if existing:
                print(f"⚠️  User {u['email']} exists - reusing")
                created_users.append(existing)
                continue

            user = User(
                name=u['name'],
                email=u['email'],
                role=u['role'],
                is_admin=u['is_admin'],
                verified=True,
                balance=u['balance'],
            )
            user.set_password(u['password'])
            db.session.add(user)
            db.session.flush()
            created_users.append(user)
            print(f"✅ Created user {u['email']}")

        db.session.commit()

        # Map affiliates (exclude admin)
        affiliates = [u for u in created_users if not u.is_admin]
        if len(affiliates) < 4:
            print("⚠️  Less than 4 affiliates found/created; aborting further seed steps.")
            return

        # 2) Create 10 offers (Products)
        products = []
        for i in range(1, 11):
            name = f"Offer {i}"
            affiliate_owner = affiliates[(i - 1) % len(affiliates)]
            existing = Product.query.filter_by(name=name, user_id=affiliate_owner.id).first()
            if existing:
                products.append(existing)
                continue

            p = Product(
                user_id=affiliate_owner.id,
                name=name,
                description=f"Deterministic seeded offer {i}",
                price=f"{9 + i}.99",
                affiliate_link=f"https://offers.example.com/{i}?ref={affiliate_owner.referral_code}",
                network='seed-net',
                category='seed',
                commission_rate='10%',
                is_active=True,
            )
            db.session.add(p)
            db.session.flush()
            products.append(p)

        db.session.commit()

        # 3) Create 50 clicks across the 10 products
        total_clicks = 50
        clicks = []
        for i in range(total_clicks):
            product = products[i % len(products)]
            # rotate user assignment; make some clicks anonymous
            user = affiliates[i % len(affiliates)] if (i % 5) != 0 else None
            existing = None
            click = AffiliateClick(
                product_id=product.id,
                user_id=user.id if user else None,
                ip_address=f"192.0.2.{(i % 250) + 1}",
            )
            db.session.add(click)
            clicks.append(click)

        db.session.commit()

        # 4) Create 12 commission/conversion records, mark 8 as approved/paid
        commissions = []
        for i in range(12):
            product = products[i % len(products)]
            # assign to one of the product owners
            owner = User.query.get(product.user_id)
            amount = Decimal('5.00') + Decimal(i)
            status = CommissionStatus.PENDING
            approved_at = None
            paid_at = None
            if i < 8:
                status = CommissionStatus.APPROVED
                approved_at = (product.created_at or None)
                # make half of approved commissions paid
                if i < 3:
                    status = CommissionStatus.PAID
                    paid_at = None

            existing = Commission.query.filter_by(user_id=owner.id, product_id=product.id, amount=amount).first()
            if existing:
                commissions.append(existing)
                continue

            c = Commission(
                user_id=owner.id,
                network='seed-net',
                product_id=product.id,
                amount=amount,
                status=status,
                external_ref=f"seed_conv_{i+1}",
            )
            if approved_at:
                c.approved_at = approved_at
            if paid_at:
                c.paid_at = paid_at

            db.session.add(c)
            db.session.flush()
            commissions.append(c)

        db.session.commit()

        # 5) Create 3 completed payouts (WithdrawalRequest) and corresponding Transactions
        payout_amounts = [Decimal('50.00'), Decimal('75.25'), Decimal('100.00')]
        for idx, amt in enumerate(payout_amounts):
            user = affiliates[idx % len(affiliates)]
            existing = WithdrawalRequest.query.filter_by(user_id=user.id, amount=amt, status=WithdrawalStatus.PAID).first()
            if existing:
                print(f"⚠️  Payout for {user.email} ${amt} exists - skipping")
                continue

            wr = WithdrawalRequest(
                user_id=user.id,
                amount=amt,
                status=WithdrawalStatus.PAID,
            )
            db.session.add(wr)
            db.session.flush()

            # also add Transaction record for payout
            tr = Transaction(
                user_id=user.id,
                type=TransactionType.WITHDRAWAL,
                amount=amt,
                status=TransactionStatus.COMPLETED,
                description=f"Seeded payout ${amt}",
            )
            db.session.add(tr)

        try:
            db.session.commit()
            print("\n✅ Deterministic seed completed successfully!\n")
            print("Summary:")
            print(f"- Users: {len(created_users)} (1 admin, 4 affiliates)")
            print(f"- Offers: {len(products)}")
            print(f"- Clicks: {AffiliateClick.query.count()}")
            print(f"- Commissions: {Commission.query.count()}")
            print(f"- Completed payouts: {WithdrawalRequest.query.filter_by(status=WithdrawalStatus.PAID).count()}")
            print()
        except Exception as e:
            db.session.rollback()
            print(f"\n❌ Error during seeding: {str(e)}\n")
            sys.exit(1)


if __name__ == '__main__':
    app = Flask(__name__)
    app.config.from_object(Config)
    db.init_app(app)

    create_deterministic_seed(app)

File: backend-node/prisma/seed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("@prisma/client");
const bcryptjs_1 = __importDefault(require("bcryptjs"));
const prisma = new client_1.PrismaClient();
async function main() {
    console.log('🌱 Running Prisma deterministic seed...');
    // 1) Users: 1 admin + 4 affiliates
    const usersSpec = [
        { name: 'Admin User', email: 'admin@lonaat.com', password: 'Admin123!', role: 'ADMIN', is_admin: true, balance: '0' },
        { name: 'Affiliate A', email: 'aff1@lonaat.com', password: 'AffPass1!', role: 'USER', is_admin: false, balance: '100' },
        { name: 'Affiliate B', email: 'aff2@lonaat.com', password: 'AffPass2!', role: 'USER', is_admin: false, balance: '200' },
        { name: 'Affiliate C', email: 'aff3@lonaat.com', password: 'AffPass3!', role: 'USER', is_admin: false, balance: '300' },
        { name: 'Affiliate D', email: 'aff4@lonaat.com', password: 'AffPass4!', role: 'USER', is_admin: false, balance: '400' },
    ];
    const createdUsers = [];
    for (const [i, u] of usersSpec.entries()) {
        const passwordHash = await bcryptjs_1.default.hash(u.password, 10);
        const referral_code = `SEED${i + 1}`;
        const existing = await prisma.user.findUnique({ where: { email: u.email } });
        if (existing) {
            createdUsers.push(existing);
            console.log(`⚠️  User ${u.email} exists - reusing`);
            continue;
        }
        const user = await prisma.user.create({
            data: {
                name: u.name,
                email: u.email,
                password: passwordHash,
                role: u.role,
                balance: u.balance,
                isActive: true,
            },
        });
        createdUsers.push(user);
        console.log(`✅ Created user ${u.email}`);
    }
    // Map affiliates (exclude admin)
    const affiliates = createdUsers.filter((x) => x.role !== 'ADMIN');
    if (affiliates.length < 4) {
        console.warn('⚠️  Less than 4 affiliates created — stopping further seed steps.');
    }
    // 2) Offers: 10 offers
    const offers = [];
    for (let i = 1; i <= 10; i++) {
        const externalOfferId = `seed_offer_${i}`;
        const seller = affiliates[(i - 1) % Math.max(affiliates.length, 1)];
        const existing = await prisma.offer.findUnique({ where: { externalOfferId } });
        if (existing) {
            offers.push(existing);
            continue;
        }
        const offer = await prisma.offer.create({
            data: {
                title: `Offer ${i}`,
                description: `Deterministic seeded offer ${i}`,
                url: `https://offers.example.com/${i}`,
                network: 'seed-net',
                externalOfferId,
                isActive: true,
                payout: `${5 + i}`,
                sellerId: seller ? seller.id : undefined,
            },
        });
        offers.push(offer);
    }
    // 3) Clicks: 50 clicks across offers
    const totalClicks = 50;
    const clicks = [];
    for (let i = 0; i < totalClicks; i++) {
        const offer = offers[i % offers.length];
        const user = i % 5 !== 0 ? affiliates[i % affiliates.length] : null; // some anonymous
        const clickId = `seed_click_${i + 1}`;
        const clickToken = `seed_token_${i + 1}`;
        const existing = await prisma.click.findUnique({ where: { clickId } });
        if (existing) {
            clicks.push(existing);
            continue;
        }
        const timeBucket = Math.floor(Date.now() / 5000)
        const click = await prisma.click.create({
            data: {
                offerId: offer.id,
                adId: offer.id,
                userId: user ? user.id : 0,
                timeBucket,
                clickId,
                clickToken,
                ip: `192.0.2.${(i % 250) + 1}`,
                userAgent: 'SeedAgent/1.0',
                user_id: user ? user.id : undefined,
            },
        });
        clicks.push(click);
    }
    // 4) Conversions: 12 conversions
    const conversions = [];
    for (let i = 0; i < 12; i++) {
        const offer = offers[i % offers.length];
        const click = clicks[i % clicks.length];
        const amount = (10 + i).toFixed(2);
        const existing = await prisma.conversion.findFirst({ where: { offerId: offer.id, amount: amount } });
        if (existing) {
            conversions.push(existing);
            continue;
        }
        const conv = await prisma.conversion.create({
            data: {
                offerId: offer.id,
                clickId: click.clickId,
                amount: amount,
                status: i < 8 ? 'completed' : 'pending',
            },
        });
        conversions.push(conv);
    }
    // 5) Commissions: 12 commissions, make 8 approved/paid-like
    const commissions = [];
    for (let i = 0; i < 12; i++) {
        const offer = offers[i % offers.length];
        const owner = offer.sellerId ? await prisma.user.findUnique({ where: { id: offer.sellerId } }) : affiliates[0];
        const click = clicks[i % clicks.length];
        const amount = (5 + i).toFixed(2);
        const external_ref = `seed_comm_${i + 1}`;
        const existing = await prisma.commission.findFirst({ where: { external_ref } });
        if (existing) {
            commissions.push(existing);
            continue;
        }
        const status = i < 8 ? 'paid' : 'pending';
        const c = await prisma.commission.create({
            data: {
                user_id: owner.id,
                click_id: click.id,
                network: 'seed-net',
                product_id: offer.id,
                amount: amount,
                status,
                external_ref,
            },
        });
        commissions.push(c);
    }
    // 6) Payouts: 3 completed payments (as payouts)
    const payoutAmounts = ['50.00', '75.25', '100.00'];
    for (let i = 0; i < payoutAmounts.length; i++) {
        const user = affiliates[i % affiliates.length];
        const transactionId = `seed_payout_${i + 1}`;
        const existing = await prisma.payment.findUnique({ where: { transactionId } });
        if (existing) {
            console.log(`⚠️  Payout ${transactionId} exists - skipping`);
            continue;
        }
        await prisma.payment.create({
            data: {
                userId: user.id,
                amount: payoutAmounts[i],
                currency: 'USD',
                status: 'completed',
                provider: 'payout',
                transactionId,
            },
        });
    }
    // Summary
    const counts = await Promise.all([
        prisma.user.count(),
        prisma.offer.count(),
        prisma.click.count(),
        prisma.conversion.count(),
        prisma.commission.count(),
        prisma.payment.count(),
    ]);
    console.log('\n✅ Prisma seed complete');
    console.log(`- Users: ${counts[0]}`);
    console.log(`- Offers: ${counts[1]}`);
    console.log(`- Clicks: ${counts[2]}`);
    console.log(`- Conversions: ${counts[3]}`);
    console.log(`- Commissions: ${counts[4]}`);
    console.log(`- Payments: ${counts[5]}`);
}
main()
    .catch((e) => {
    console.error(e);
    process.exit(1);
})
    .finally(async () => {
    await prisma.$disconnect();
});
